### ISSUE TITLE
generix collections in Ruby land

### ISSUE THREAD

**Issue by kares** [jruby/jruby#3795]
suppose a collection (or any other generic type) such as `java.util.ArrayList` never leaves Ruby-land.

such as presented here: https://github.com/jruby/jruby/issues/2049#issuecomment-208690695 (implicit conversion leading to user confusion on times)

in cases such as those it might not be desirable for type conversions to be happening whenever a `java.lang.Object` boundary is seen (e.g. `list.add Time.now`). but instead of tricky escape-analysis a reasonable way this could work is if target type information was available. some quick ideas : 
```ruby
list = java.util.ArrayList.of(org.jruby.RubyTime)
list.add Time.now # no conversion

list = java.util.LinkedList.new
list.typed! 'RubyString'

set = java.util.HashSet['java.lang.Integer'].new 
# true types Set<Integer> - could work for any generixized type

# the last example might become very real with Java 9 around
# ... since we would likely want to be able to : 
set = java.util.HashSet[:int].new # Set<int>
```
consequences of such wrapped Java types is that their need to become `__persistent__` to store additional generic type information. except maybe on Java 9 where there likely is to be a way to tell a `Set<Integer>` from a `Set<int>`

**Comment by enebo** [comment]
@kares fwiw I don't think we will ever escape that coercion away.  I am also not even sure even if we could see something escape we would anyways.  Once you convert it is really a different object at that point.

I like the spirit of this enhancement.  People who need the power will be willing to type their collections in such a way to prevent coercion.  Most people will not care and will not do this but that is fine.  This just gives us another level of control in how we interact with Java from Ruby.

**Comment by kares** [comment]
yeah, the escape was mostly a note for people expecting that if its all in Ruby it should behave differently.

like the typed collection `java.util.HashSet[:int]` syntax although it would only allow for typing of those created from within Ruby. not sure I like `set.typed! :int` or any similar method. do you think we would need to support externaly created "typed" collections? hoping its really not that important.

also wonder whether that syntax should disable any conversion e.g. `java.util.ArrayList[nil].new` ?

**Comment by enebo** [comment]
Makes me wonder if we can just make a class method to return an uncoercing instance.  At least for non-primitive types: `java.util.ArrayList.new!`.

Is there value to enforcing the type?  I think primitives takes down a whole different rabbit hole...

**Comment by kares** [comment]
> Makes me wonder if we can just make a class method to return an uncoercing instance. At least for non-primitive types: java.util.ArrayList.new!.

sounds like a simple good start and if it is to be implemented it would likely work on all Java proxies, yay :)
although it probably gets internally a little ugly to check on all those places whether toJava should happen.

**Comment by headius** [comment]
Note: Java 9 does not include any generic reification, nor will the next couple years of updates (most likely).
